src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Services/SemanticFacts/CSharpSemanticFacts.cs

```diff
@@ -236,6 +236,17 @@ public bool IsPartial(ITypeSymbol typeSymbol, CancellationToken cancellationToke
             return syntaxRefs.Any(n => ((BaseTypeDeclarationSyntax)n.GetSyntax(cancellationToken)).Modifiers.Any(SyntaxKind.PartialKeyword));
         }
 
+        public bool IsNullChecked(IParameterSymbol parameterSymbol, CancellationToken cancellationToken)
+        {
+            var syntaxRefs = parameterSymbol.DeclaringSyntaxReferences;
+            // We never expect multiple declarations for a parameter symbol.
+            // Partial method parameters are distinct symbols on each method declaration
+            Contract.ThrowIfTrue(syntaxRefs.Length > 1);
+            return syntaxRefs.SingleOrDefault() is SyntaxReference syntaxReference
```

Comment:
SingleOrDefault will throw when there are multiple references.  taht woudl be bad.

Reply:
I did this because I really do expect a parameter to never have multiple syntax references. Is there a different way of handling this scenario you would prefer? For example, just take the first reference if any, check if any of the references have the `!!` token, etc.

Reply:
I'd assume there's more than one if the containing method is partial?

Reply:
> Is there a different way of handling this scenario you would prefer? For example, just take the first reference if any, check if any of the references have the !! token, etc.

Checking if any references have `!!` seems reasonable to me.

Reply:
I think it's important to know if in any case (even in future), we can have multiple syntax locations for a parameter. I'd add a `Debug.Assert` in the base parameter symbol with a comment explaining that this API is assuming a single syntax reference.

Reply:
> I think it's important to know if in any case (even in future), we can have multiple syntax locations for a parameter. I'd add a `Debug.Assert` in the base parameter symbol with a comment explaining that this API is assuming a single syntax reference, or no references at all.

This was my inclination as well. However, I think `Debug.Assert` will realistically only end up catching the issue if an editor feature test is added which exercises such a new parameter symbol while also using this new API.

Reply:
@RikkiGibson I mean adding the assert in ParameterSymbol constructor

Reply:
Sorry, I overlooked what you actually said. If we can identify a reasonable central place to assert this assumption, I would be fine with taking it in a separate PR.